# Data Cleaning Project - Evaluation and Cleaning of E-commerce Sales Data

## Analysis Objective

To formulate more effective marketing strategies to increase revenue. The purpose of this data analysis is to mine best-selling products based on market sales data. 

The goal of this practical project is to practice assessing the cleanliness and tidiness of the data, and based on the assessment results, to cleanse the data, thereby obtaining data suitable for further analysis.

## Introduction

The original dataset records all transactions of a British online retail company from December 1, 2010, to December 9, 2011, covering the company's business data in different countries and regions around the world. 

The company mainly sells gifts for various occasions, including but not limited to birthday gifts, wedding commemoratives, Christmas gifts, etc. 

The customer base of the company mainly includes wholesalers and individual consumers, with wholesalers accounting for a significant proportion.

The meaning of each column in the data is as follows:
- `InvoiceNo`: Invoice number. A 6-digit number serving as the unique identifier for the transaction. If this code starts with the letter "c", it indicates that the transaction has been canceled.

- `StockCode`: Product code. A 5-digit number serving as the unique identifier for the product.

- `Description`: Product name.

- `Quantity`: The number of products in the transaction.

- `InvoiceDate`: Invoice date and time. The date and time when the transaction occurred.

- `UnitPrice`: Unit price. The price is in British pounds (GBP).

- `CustomerID`: Customer number. A 5-digit number serving as the unique identifier for the customer.

- `Country`: Country name. The name of the country where the customer resides.


In [1]:
import pandas as pd

## Data Reading

In [2]:
original_data = pd.read_csv("e_commerce.csv")

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## Data Assessment

In this section, I will evaluate the data contained in the DataFrame established in the previous part, referred to as "original data."

The assessment is conducted primarily from two perspectives: structure and content, which pertain to neatness and cleanliness.

Structural issues in the data refer to non-compliance with the three standards of "each column is a variable, each row is an observation, and each cell is a value."

Content issues in the data include missing data, duplicate data, and invalid data.

### Assessing Data Neatness

In [4]:
original_data.sample(10)

InvoiceNo StockCode                      Description  Quantity  \
441474    574575     84879    ASSORTED COLOUR BIRD ORNAMENT         8   
335758    566291     21755         LOVE BUILDING BLOCK WORD         2   
441021    574559     23582     VINTAGE DOILY JUMBO BAG RED          1   
70265     542000     21212  PACK OF 72 RETROSPOT CAKE CASES        24   
143683    548711     21899          KEY FOB , GARAGE DESIGN         1   
82668     543246     22050      PINK PAISLEY ROSE GIFT WRAP        25   
109857    545661     22423         REGENCY CAKESTAND 3 TIER         2   
220401    556189     22383           LUNCH BAG SUKI DESIGN         10   
54579     540957     21212  PACK OF 72 RETROSPOT CAKE CASES         3   
453457    575481     84006        MAGIC TREE -PAPER FLOWERS         6   

            InvoiceDate  UnitPrice  CustomerID         Country  
441474  11/4/2011 17:03       1.69     14971.0  United Kingdom  
335758  9/11/2011 15:31       6.25     12965.0  United Kingdom  
441021  11/4/2011 15:50       2.08     14056.0  United Kingdom  
70265   1/25/2011 10:49       0.55     14257.0  United Kingdom  
143683   4/3/2011 14:24       0.65     18116.0  United Kingdom  
82668    2/4/2011 14:14       0.42     17315.0  United Kingdom  
109857   3/4/2011 14:41      12.75     15358.0  United Kingdom  
220401   6/9/2011 11:57       1.65     15615.0  United Kingdom  
54579   1/12/2011 13:30       0.55     15311.0  United Kingdom  
453457  11/9/2011 16:28       0.85     15679.0  United Kingdom

Conclusion:
- Upon reviewing a sample of 10 rows of data, it is observed that the data adheres to the principle that "each column represents a variable, each row represents an observation, and each cell contains a value."

- Specifically, each row pertains to a transaction of a particular product, and each column represents the various variables associated with the transaction; hence, there are no structural issues.

### Data Cleanliness Evaluation 

In [ ]:
original_data.info()

# Missing Values: There are gaps in the data as the total number of rows is less than 541,909.
# 2   Description  540455 non-null  object 
# 6   CustomerID   406829 non-null  float64

# The Invoice Date should be of a datetime data type instead of a string.
# 4   InvoiceDate  541909 non-null  object

# The Customer ID should be changed to a string data type.
# 6   CustomerID   406829 non-null  float64

Conclusion:
- Based on the output, the dataset contains 541,909 observations, with missing values present in the Description and CustomerID variables.

- The InvoiceDate should be of a date data type.

- The CustomerID should be of a string data type, and data type conversion should be performed.

## Data Cleansing

Types of potential dirty data that may exist:
- Missing Data
- Duplicate Data
- Inconsistent Data
- Invalid Data
- Erroneous Data


### Assessing Missing Data

In [7]:
# Filter rows with missing values in the Description column.
original_data[original_data['Description'].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

Conclusion
- There are 1,454 transaction records missing the Description variable value.

- From the output results, these transactions missing the Description have a Unitprice of 0.

- To verify the assumption, we add additional filtering criteria to see if there are any transactions missing the Description where the Unitprice is not 0.

In [8]:
# Filter rows with missing values in the Description column,
# among which the rows where the unitprice column is not zero.

original_data[(original_data['Description'].isnull()) 
              & (original_data['UnitPrice']!=0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

Conclusion:
- The filtered result count is 0, indicating that rows with missing Description values also lack valid Unitprice values.

- Description represents the product name, and Unitprice represents the product's unit price, both of which are important variables for subsequent product transaction analysis.

- If they are both missing/invalid, we consider the data to be devoid of meaningful information, and therefore, these rows can be deleted subsequently.

- The CustomerID variable also has missing values, so missing observations should also be extracted based on the condition.

In [9]:
# Filter rows with missing values in the 'Description' column,
# among which the rows where the 'UnitPrice' column is not zero.
original_data[original_data['CustomerID'].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

Conclusion:

- CustomerID represents the customer number and is not a necessary variable for analyzing best-selling products.

- Furthermore, from the output results, it can be seen that some transaction data with missing CustomerID are still valid,

- Therefore, observations with this variable being empty are retained.

#### Assessing Duplicate Data

Based on the meanings of the data variables, although InvoiceNo, StockCode, and CustomerID are all unique identifiers,

a single transaction may include multiple items, so duplicates in InvoiceNo are permissible,

different transactions may include the same item, so duplicates in StockCode are permissible,

and customers can conduct multiple transactions or order multiple items, so duplicates in CustomerID are permissible.

Therefore, for this dataset, there is no need to assess duplicate data.

#### Assessing Inconsistent Data

Inconsistent data may exist in the country variable,

we need to check if there are multiple different values representing the same country.

In [10]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


Conclusion:

- Upon reviewing the values of the Country variable,

- "USA" and "United States" both represent the United States,

- "United Kingdom", "UK", and "U.K." all represent the United Kingdom,

- Therefore, these values should be standardized in the cleaning step, retaining only one representative value.


#### Assessing Invalid or Erroneous Data

Quickly gaining an understanding of numerical statistical information
Can be achieved through the describe method of the DataFrame.

In [11]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

Conclusion

- From the output results, there are negative values in Quantity and Unitprice, which will affect subsequent numerical analysis.

- Therefore, we first filter out the observations where the Quantity value is negative, to further assess their meaning.

In [12]:
original_data[original_data["Quantity"] < 0 ]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

Conclusion:

- From the filtered results, it appears that observations with negative values in the Quantity variable all have InvoiceNo starting with "C," which indicates that the order has been canceled.

- However, this assumption may not be accurate.

- To validate the hypothesis, we add additional filtering criteria to check for the existence of observations with negative Quantity values where InvoiceNo does not start with "C".


In [14]:
original_data[(original_data["Quantity"] < 0) 
              & (original_data["InvoiceNo"].str[0]!="C") ]
               # Rows where the product quantity is negative, but the invoice number does not start with "C".

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

Conclusion:

- The above hypothesis is proven incorrect, as there still exist observations with negative values in the Quantity variable, but the InvoiceNo does not start with "C."

- However, according to the above output results, the Unitprice for these filtered observations is all 0, therefore, we will add the condition of Unitprice to the filter.


In [15]:
original_data[(original_data["Quantity"] < 0)
              &(original_data["InvoiceNo"].str[0] != "C")
              &(original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

Conclusion:

- Based on the output results, the hypothesis is confirmed when the `Quantity` variable is negative, the observations meet one of the following conditions:
    - `InvoiceNo` starts with "C," indicating that the order has been canceled.
    
    - `UnitPrice` is 0, indicating that the unit price is zero British pounds.

- These transaction records are not valid sales data, do not contribute to sales, and are not within the scope of subsequent analysis.

- Therefore, we will delete the observations with a negative `Quantity` in the data cleaning step.


In [16]:
# Filter rows where the column 'UnitPrice' is also negative.
original_data[original_data["UnitPrice"]<0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

Conclusion:

Adjust bad debt refers to adjustments for uncollectible debts, which do not belong to actual product transaction data; therefore, 

it will also be deleted during the data cleaning process.


## Data Cleaning


Based on the conclusions drawn from the previous evaluation section, the data cleaning we need to carry out includes:

- Convert the data type of the 'InvoiceDate' variable to a date and time

- Convert the data type of the 'Customer ID' variable to a string

- Delete the missing observation values of the 'Description' variable

- Replace the 'Country' variable value 'USA' with 'United States'`

- Replace the 'Country' variable values' UK 'and' UK 'with' United Kingdom '`

- Delete observations with negative values for the 'Quantity' variable

- Delete observations with negative values for the 'UnitPrice' variable

Create a new variable 'cleaned_data' to save the cleaned data

In [17]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [19]:
# Convert the data type of the 'InvoiceDate' variable to a date and time
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

In [20]:
# Convert the data type of the 'Customer ID' variable to a string
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

In [23]:
# Delete the 0 after the decimal place of the 'Customer ID' value
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0,-2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

In [26]:
# Delete the missing observation values of the 'Description' variable
cleaned_data = cleaned_data.dropna(subset=["Description"])
cleaned_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[540455 rows x 8 columns]

In [27]:
# Check if there are still empty values in the Description column
cleaned_data['Description'].isnull().sum()

0

In [28]:
# Replace the 'Country' variable value 'USA' with 'United States'`
cleaned_data["Country"] = cleaned_data["Country"].replace({'USA':'United States'})

In [30]:
cleaned_data["Country"] == "USA"

0         False
1         False
2         False
3         False
4         False
          ...  
541904    False
541905    False
541906    False
541907    False
541908    False
Name: Country, Length: 540455, dtype: bool

In [29]:
# Check if there are any rows that have not been modified to 'United States'

len(cleaned_data[cleaned_data["Country"] == "USA"])

0

In [37]:
# Replace the 'Country' variable values 'UK' and 'UK' with 'United Kingdom'

cleaned_data["Country"] = cleaned_data["Country"].replace({'UK':'United Kingdom',"U.K.":"United Kingdow"})

print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0


In [38]:
# Delete observations with negative values for the 'Quantity' variable
# Save non negative values to clean_data
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [40]:
# Check if there are any rows less than 0 in cleaned_data
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

In [41]:
# Delete observations with negative values for the 'UnitPrice' variable
cleaned_data = cleaned_data[cleaned_data['UnitPrice'] >= 0]

In [43]:
# Check if there are still negative rows in the UnitPrice column
len(cleaned_data[cleaned_data['UnitPrice'] < 0])

0

## Save the cleaned data

After completing data cleaning, save the clean and tidy data to a new file named 'e_commerce_clean. csv'.

In [44]:
cleaned_data.to_csv('e_commerce_cleaned.csv',index = False)

In [45]:
pd.read_csv('e_commerce_cleaned.csv').head()

/var/folders/55/wz8gkxh90xz7m0yw6vjysd680000gn/T/ipykernel_81290/563775235.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv('e_commerce_cleaned.csv').head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom